File: ruby/perfect-numbers/_iterations/1/perfect_numbers.rb
module PerfectNumber
  def self.classify(n)
    raise RuntimeError, "n must be positive" unless n.positive?
    
    aliquot_sum = factors(n).reject { |x| x == n }.sum

    return "perfect" if aliquot_sum == n
    return "abundant" if aliquot_sum > n
    "deficient"
  end

  def self.factors(n)
    (1..Math.sqrt(n)).filter { |d| n % d == 0 }.flat_map { |f| [f, n / f] }.uniq
  end
end

File: clojure/perfect-numbers/_iterations/1/src/perfect_numbers.clj
(ns perfect-numbers)

(defn- factor-for? [n factor]
  (zero? (mod n factor)))

(defn- factors [n]
  (filter (partial factor-for? n) (range 1 n)))

(defn- aliquot-sum [n]
  (reduce + (factors n)))

(defn classify [n]
  (when-not (pos? n)
    (throw (IllegalArgumentException. "n must be positive")))
  (let [aliquot (aliquot-sum n)]
    (cond
      (= aliquot n) :perfect
      (> aliquot n) :abundant
      :else         :deficient)))
